### DISCUSSION TITLE
Karpenter support

### DISCUSSION THREAD

**Discussion by drmorr0** [acrlabs/simkube#71] [Feature Requests and Future Work]
Right now we interact with Cluster Autoscaler via an externalgrpc cloud provider.  It would be really great if we can also support Karpenter, since there is a lot of interest and excitement about this project right now.  If we end up moving to KWOK (#70) we will just write a KWOK provider for karpenter, otherwise we might have to build something inside SimKube itself.  This is a tracking issue for this project.

**Comment by Huang-Wei** [comment]
Awesome! would love to see it come into play.

cc @wzshiming

**Comment by wzshiming** [comment]
Awesome!

FYI: It looks like karpenter has been using kwok for [performance testing](https://github.com/search?q=repo%3Aaws%2Fkarpenter-core+kwok&type=pullrequests).

**Reply by drmorr0** [reply]
The link has moved now that karpenter is a CNCF project: https://github.com/kubernetes-sigs/karpenter/tree/main/kwok

**Comment by drmorr0** [comment]
The kwok provider included with Karpenter only supports a hard-coded list of node types that don't actually correspond to any "real" node types.  If we want to run a scheduling simulation with Karpenter in the same way that we do with Cluster Autoscaler, we somehow need to tell Karpenter what the "real" list of node types it can draw from is... One idea would be to just make Karpenter read from the `kwok-provider-templates` configmap that cluster autoscaler does, but I feel like this limits/doesn't provide a super fair comparison for Karpenter, since one of its selling points is being able to draw from _any_ instance type that's available.  It might be a good starting point, though.

Another point to consider (this _could_ be an issue with Cluster Autoscaler, too, but probably more-so with Karpenter) is that SimKube and/or KWOK and/or the autoscaler needs to know something about the "simulated" availability of an instance type.  Ideally we'd like to be able to simulate scenarios where the "best" or "desired" instance type is unavailable so that we can see what our autoscaler does in response.  I'm not 100% sure how to implement this yet.

**Reply by drmorr0** [reply]
I'm going to attend the CA/Karpenter meeting on Monday Feb 4 to talk about this a bit more (WG agenda [here](https://docs.google.com/document/d/18BT0AIMugpNpiSPJNlcAL2rv69yAE6Z06gUVj7v_clg/edit#heading=h.pumnznxk0bjt))

**Reply by drmorr0** [reply]
And here's an issue I opened in the Karpenter repo: https://github.com/kubernetes-sigs/karpenter/issues/987

**Comment by drmorr0** [comment]
The karpenter project now lets you define custom instance types for the KWOK provider, going to close this out.
